README.md
cmdMQ - README
====================

+ Name: cmdMQ

+ Version: 1.0.0

+ Release date: 2013-03-21

+ Author: Nicola Asuni

+ Copyright (2013-2013):

> > Fubra Limited  
> > Manor Coach House  
> > Church Hill  
> > Aldershot  
> > Hampshire  
> > GU12 4RQ  
> > <http://www.fubra.com>  
> > <support@fubra.com>  

+ Website: https://github.com/fubralimited/cmdmq


SOFTWARE LICENSE:
-----------------

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

See LICENSE.TXT file for more information.


DESCRIPTION:
------------

cmdMQ is a Free Open Source Software system to send commands to a remote computer using RabbitMQ queues.
This system uses an RPC (Remote Procedure Call) model to send commands from one computer node to another and get the results as JSON formatted data.
The RabbitMQ server (broker) can be installed on the sender computer node, the receiver computer node or another computer node.
The cmdMQ-Receiver runs and execute received commands as root. The only commands executed are the one specified on the configuration file using a regular expression syntax.


	+--------------+      rpc_queue      +----------------+ 
	|              |--->[ | | | | | ]--->|                |
	| cmdMQ-Sender |                     | cmdMQ-Receiver |
	|              |<---[ | | | | | ]<---|                |
	+--------------+     reply_queue     +----------------+


## cmdmq_sender ##

This section contains the software to be installed on the computer that sends commands.
Once configured, you can send a command to the queue using the following systax:

	cmdmq_sender.py -e "<command to be executed>"

For example:

	cmdmq_sender.py -e "ls -la"
	
This file must be marked as executable:

	chmod +x cmdmq_sender.py
	
Or called using python:

	python cmdmq_sender.py
		

This command returns a JSON object with the following information:
		
	{
		'ret':<ERROR_CODE>,
		'out':<COMMAND_OUTPUT>,
		'err':<ERROR_MESSAGE>
	}

## cmdmq_receiver ##

This section contains the software to be installed on the computer that receives and execute commands.
Once configured, this program can be started using one of the provided init scripts or manually:

	cmdmq_receiver.py



## INSTALLATION ##

### Install RabbitMQ on CentOS ###

Install EPEL, then:

	 yum -y install openssl
	 yum -y install erlang

Download the RabbitMQ RPM

	wget http://www.rabbitmq.com/releases/rabbitmq-server/v3.0.4/rabbitmq-server-3.0.4-1.noarch.rpm

Install RabbitMQ:

	 rpm --import http://www.rabbitmq.com/rabbitmq-signing-key-public.asc
	 yum -y install rabbitmq-server-3.0.4-1.noarch.rpm

Edit the RabbitMQ main configuration file only if you want to use SSL

	nano /etc/rabbitmq/rabbitmq.config

	 [
	   {rabbit, [
		  {default_user, <<"catnrmq">>},
		  {default_pass, <<"N8IJYBxbCnhqWZ8DshvTXofTONtegRzZ">>},
		  {ssl_listeners, [5671]},
		  {ssl_options, [{cacertfile,"/etc/rabbitmq/ssl/cacert.pem"},
						 {certfile,"/etc/rabbitmq/ssl/cert.pem"},
						 {keyfile,"/etc/rabbitmq/ssl/key.pem"},
						 {verify,verify_none},
						 {fail_if_no_peer_cert,false}]}
		]}
	 ].

Copy the certificate files as indicated on the configuration file.
For the generation of the SSL certificates please consult:
http://www.rabbitmq.com/ssl.html

Start the service:

	service rabbitmq-server start

Start service at reboot:

	chkconfig rabbitmq-server on.

Use rabbitmqctl stop to stop the server.
Use rabbitmqctl status to check whether it is running.


### Install RabbitMQ on NAS4Free ###

Create the user "rabbitmq" with shell "nologin" and primary group "daemon".

To install additional packages we need to set the correct repository (discarded after reboot):

	setenv PACKAGESITE ftp://ftp.freebsd.org/pub/FreeBSD/ports/amd64/packages-9-current/Latest/

Install the following packages (used for the mesaging system):

	pkg_add -r openssl
	cp /usr/local/openssl/openssl.cnf.sample /usr/local/openssl/openssl.cnf
 
	pkg_add -r rabbitmq

Edit the /etc/rc.conf file and add the following lines:

	 rabbitmq_enable="YES"
	 rabbitmq_user="root"

Edit the RabbitMQ main configuration file only if you want to use SSL

 nano /etc/rabbitmq/rabbitmq.config

	 [
	   {rabbit, [
		  {default_user, <<"catnrmq">>},
		  {default_pass, <<"N8IJYBxbCnhqWZ8DshvTXofTONtegRzZ">>},
		  {ssl_listeners, [5671]},
		  {ssl_options, [{cacertfile,"/etc/rabbitmq/ssl/cacert.pem"},
						 {certfile,"/etc/rabbitmq/ssl/cert.pem"},
						 {keyfile,"/etc/rabbitmq/ssl/key.pem"},
						 {verify,verify_none},
						 {fail_if_no_peer_cert,false}]}
		]}
	 ].


Copy the configuration file:

	cp /etc/rabbitmq/rabbitmq.config /usr/local/etc/rabbitmq/rabbitmq.config 

Copy the certificate files as indicated on the configuration file.

To start the RabbitMQ server:
 
	/usr/local/etc/rc.d/rabbitmq start

### Reset RabbitMQ ###

To reset the RabbitMQ user/password:

	 rabbitmqctl stop_app
	 rabbitmqctl force_reset
	 rabbitmqctl start_app

### SSL TEST ###

In one terminal window execute the following command: 

	openssl s_server -accept 8443 -cert /etc/rabbitmq/server/cert.pem -key /etc/rabbitmq/server/key.pem -CAfile /etc/rabbitmq/testca/cacert.pem

In another terminal window execute

	openssl s_client -connect 127.0.0.1:8443 -cert /etc/rabbitmq/client/cert.pem -key /etc/rabbitmq/client/key.pem -CAfile /etc/rabbitmq/testca/cacert.pem

If the certificates and keys have been correctly created, an SSL connection establishment sequence will appear and the terminals will be linked.
Input from either terminal will appear on the other.


### Install cmdMQ on CentOS ###

Install python packages

	yum -y install python python-pip python-pika

Copy the cmdmq_receiver.py script to /usr/local/sbin/cmdmq_receiver.py and mark it as executable with
 
	chmod +x /usr/local/sbin/cmdmq_receiver.py

Copy the default.conf configuration file to /etc/cmdmq_receiver/default.conf and edit it.

Copy the cmdmq_receiver init script to /etc/init.d/cmdmq_receiver and mark it as executable with 

	chmod +x /etc/init.d/cmdmq_receiver


### Install cmdMQ on NAS4Free ###

To install additional packages we need to set the correct repository (discarded after reboot):

	setenv PACKAGESITE ftp://ftp.freebsd.org/pub/FreeBSD/ports/amd64/packages-9-current/Latest/

install the following packages (used for the mesaging system):

	pkg_add -r python
	pkg_add -r py27-utils
	pkg_add -r py27-pip
 
 reboot

	pip install http://pypi.python.org/packages/source/p/pika/pika-0.9.9.tar.gz

Copy the cmdmq_receiver.py script to /usr/local/sbin/cmdmq_receiver.py and mark it as executable with
 
	chmod +x /usr/local/sbin/cmdmq_receiver.py

Copy the default.conf configuration file to /etc/cmdmq_receiver/default.conf and edit it.

Copy the cmdmq_receiver_fsb init script to /usr/local/etc/rc.d/cmdmq_receiver and mark it as executable with 

	chmod +x /usr/local/etc/rc.d/cmdmq_receiver

Edit the /etc/rc.conf file and add the following lines to start the service at boot:

	cmdmq_receiver_enable="YES"


CHANGELOG.TXT
1.0.0 (2013-03-21)
	- First public release.

0.0.0 (2013-03-06)
	- project start.


receiver/cmdmq_receiver.py
#!/usr/bin/env python

#==============================================================================+
# File name   : cmdmq_receiver.py
# Begin       : 2013-03-06
# Last Update : 2013-03-21
# Version     : 1.0.0
#
# Description : RabbitMQ RPC server used to receive and execute ZFS commands.
#
# Installation: Copy this script on the NAS4Free node.
#               Set the execuable permission (chmod +x cmdmq_receiver.py).
#               Start the RabbitMQ server.
#               Execute this script.
#
#     Website : https://github.com/fubralimited/cmdmq
#
# Author: Nicola Asuni
#
# (c) Copyright:
#               Fubra Limited
#               Manor Coach House
#               Church Hill
#               Aldershot
#               Hampshire
#               GU12 4RQ
#               UK
#               http://www.fubra.com
#               support@fubra.com
#
# License:
#    Copyright (C) 2013-2013 Fubra Limited
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as
#    published by the Free Software Foundation, either version 3 of the
#    License, or (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    See LICENSE.TXT file for more information.
#==============================================================================+

import pika
import subprocess
import json
import time
import logging
import re
import ssl
import os
import sys
import argparse
from ConfigParser import SafeConfigParser

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

# This script must be executed by root
if not os.geteuid()==0:
	sys.exit("\n*** Only root can run this script ***\n")

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# PROCESS THE COMMAND LINE ARGUMENTS

parser = argparse.ArgumentParser(description='Listen to a RabbitMQ queue for commands to be executed')
parser.add_argument('-c', '--config', action='store', nargs='?', const='cmdmq_receiver.conf', default='/etc/cmdmq_receiver/default.conf', type=file, required=False, help='configuration file - by default /etc/cmdmq_receiver/default.conf, or local cmdmq_receiver.conf if only -c is specified', dest='config_file')
args = parser.parse_args()

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# CONFIGURATION PARAMETERS

parser = SafeConfigParser()
parser.readfp(args.config_file)

# RabbitMQ server IP address
rmq_ip = parser.get('general', 'rmq_ip')

# RabbitMQ server port number
rmq_port = parser.getint('general', 'rmq_port')

# queue name (storage node hostname)
rmq_queue = parser.get('general', 'rmq_queue')

# RabbitMQ authentication
rmq_username = parser.get('general', 'rmq_username')
rmq_password = parser.get('general', 'rmq_password')

# SSL options
rmq_ssl = parser.getboolean('general', 'rmq_ssl')
rmq_ssl_options = parser.get('general', 'rmq_ssl_options')

# number of seconds to wait for restart after general error
retry_time = parser.getint('general', 'retry_time')

# log file name
log_file_name = parser.get('general', 'log_file')

# log file
log_file = open(log_file_name, "a")

# create a list of patterns for allowed commands
allowed_cmds = parser.get('filter', 'allowed_cmds').split(',')

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

# redirect stdout to a log file
old_stdout = sys.stdout
sys.stdout = log_file

# log only critical error messages
logging.basicConfig(filename=log_file_name, level=logging.ERROR)

# function to print a log message on stdout
def log_message(message):
	now = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime())
	print "%s - %s" % (now, message)
	sys.stdout.flush()

log_message('*** - PROGRAM STARTED - trying to connect')

while True:
	try:
		conn_credentials = pika.PlainCredentials(rmq_username, rmq_password)
		conn_params = pika.ConnectionParameters(
			host=rmq_ip,
			port=rmq_port,
			virtual_host=None,
			credentials=conn_credentials,
			channel_max=None,
			frame_max=None,
			heartbeat_interval=None,
			ssl=rmq_ssl,
			ssl_options=rmq_ssl_options,
			connection_attempts=5,
			retry_delay=2,
			socket_timeout=None,
			locale=None,
			backpressure_detection=None)

		connection = pika.BlockingConnection(conn_params)
		channel = connection.channel()
		channel.queue_declare(queue=rmq_queue)

		def on_request(ch, method, props, body):

			# acknowledge command receipt
			ch.basic_ack(delivery_tag=method.delivery_tag)

			log_message('[R] - Received command: ' + body)

			# check for invalid characters
			match = re.search(r'[&><|`]', body)
			if match is not None:
				# return error message
				log_message('[E] - Invalid character on command: ' + body)
				retval = json.dumps({'ret':1, 'out':'', 'err':'invalid character on command'})
			else:
				# initialize empty command
				cmd = ''
				# filter allowed commands
				for recmd in allowed_cmds:
					p = re.compile(recmd)
					match = p.search(body)
					if match is not None:
						# command is valid, exit the loop
						cmd = body
						break

				if cmd:
					# execute the requested command
					process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
					# wait for the process to terminate and store stdout, stderr and error code
					out, err = process.communicate()
					errcode = process.returncode
					# encode answer as json object
					retval = json.dumps({'ret':errcode, 'out':out, 'err':err})
					log_message('[X] - Command executed: ' + cmd)
				else:
					# return error message
					retval = json.dumps({'ret':1, 'out':'', 'err':'unrecognized command'})
					log_message('[E] - Unrecognized command: ' + body)

			# send results back
			ch.basic_publish(exchange='', routing_key=props.reply_to, properties=pika.BasicProperties(correlation_id=props.correlation_id), body=retval)

		channel.basic_qos(prefetch_count=1)
		channel.basic_consume(on_request, queue=rmq_queue)

		log_message('[S] - Awaiting remote command requests')

		channel.start_consuming()

	except KeyboardInterrupt:
		try:
			channel.stop_consuming()
			connection.close()
		except:
			pass
		log_message('[E] - Program manually interrupted')
		sys.stdout = old_stdout
		log_file.close()
		exit(1)
	except Exception,e:
		try:
			# retry to connect
			log_message('[E] - ' + str(e) + ' - restart in ' + str(retry_time) + ' secs')
			try:
				channel.stop_consuming()
				connection.close()
			except:
				pass
			time.sleep(retry_time)
		except KeyboardInterrupt:
			try:
				channel.stop_consuming()
				connection.close()
			except:
				pass
			log_message('[E] - Program manually interrupted')
			sys.stdout = old_stdout
			log_file.close()
			exit(1)

#==============================================================================+
# END OF FILE
#==============================================================================+


sender/cmdmq_sender.py
#!/usr/bin/env python

#==============================================================================+
# File name   : cmdmq_sender.py
# Begin       : 2013-03-06
# Last Update : 2013-03-21
# Version     : 1.0.0
#
# Description : RabbitMQ RPC client used to send commands to a remote
#               receiver.
#
#     Website : https://github.com/fubralimited/cmdmq
#
# Installation: Copy this script on the client node.
#               Set the execuable permission (chmod +x cmdmq_sender.py).
#               Call this script with the command to send as arguments.
#               For example: cmdmq_sender.py -e "ls -lah"
#
# Author: Nicola Asuni
#
# (c) Copyright:
#               Fubra Limited
#               Manor Coach House
#               Church Hill
#               Aldershot
#               Hampshire
#               GU12 4RQ
#               UK
#               http://www.fubra.com
#               support@fubra.com
#
# License:
#    Copyright (C) 2013-2013 Fubra Limited
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as
#    published by the Free Software Foundation, either version 3 of the
#    License, or (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    See LICENSE.TXT file for more information.
#==============================================================================+

import pika
import uuid
import json
import logging
import ssl
import sys
import argparse
from ConfigParser import SafeConfigParser

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
# PROCESS THE COMMAND LINE ARGUMENTS

parser = argparse.ArgumentParser(description='Send a command to a remote receiver using RabbitMQ queue')
parser.add_argument('-c', '--config', action='store', nargs='?', const='cmdmq_sender.conf', default='/etc/cmdmq_sender/default.conf', type=file, required=False, help='configuration file - by default /etc/cmdmq_sender/default.conf, or local cmdmq_sender.conf if only -c is specified', dest='config_file')
parser.add_argument('-e', '--command', action='store', required=True, help='command to be sent enclosed by single or double quotes', dest='command')
args = parser.parse_args()

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# CONFIGURATION PARAMETERS

parser = SafeConfigParser()
parser.readfp(args.config_file)

# RabbitMQ server IP address
rmq_ip = parser.get('general', 'rmq_ip')

# RabbitMQ server port number
rmq_port = parser.getint('general', 'rmq_port')

# queue name (storage node hostname)
rmq_queue = parser.get('general', 'rmq_queue')

# RabbitMQ authentication
rmq_username = parser.get('general', 'rmq_username')
rmq_password = parser.get('general', 'rmq_password')

# SSL options
rmq_ssl = parser.getboolean('general', 'rmq_ssl')
rmq_ssl_options = parser.get('general', 'rmq_ssl_options')

# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

# discard error messages
logging.basicConfig(filename='/dev/null', level=logging.CRITICAL)

class cmdRpcClient(object):

    def __init__(self):
		try:
			conn_credentials = pika.PlainCredentials(rmq_username, rmq_password)
			conn_params = pika.ConnectionParameters(
				host=rmq_ip,
				port=rmq_port,
				virtual_host=None,
				credentials=conn_credentials,
				channel_max=None,
				frame_max=None,
				heartbeat_interval=None,
				ssl=rmq_ssl,
				ssl_options=rmq_ssl_options,
				connection_attempts=5,
				retry_delay=2,
				socket_timeout=None,
				locale=None,
				backpressure_detection=None)

			self.connection = pika.BlockingConnection(conn_params)
			self.channel = self.connection.channel()
			result = self.channel.queue_declare(exclusive=True)
			self.callback_queue = result.method.queue
			self.channel.basic_consume(self.on_response, no_ack=True, queue=self.callback_queue)
		except:
			# print error message and exit
			print json.dumps({'ret':1, 'out':'', 'err':'connection error'})
			exit(1)

    def on_response(self, ch, method, props, body):
        if self.corr_id == props.correlation_id:
            self.response = body

    def call(self, cmd):
        self.response = None
        self.corr_id = str(uuid.uuid4())
        self.channel.basic_publish(exchange='',
					routing_key=rmq_queue,
					properties=pika.BasicProperties(
						reply_to = self.callback_queue,
						correlation_id = self.corr_id,
						),
					body=str(cmd))
        while self.response is None:
            self.connection.process_data_events()
        return self.response

command_rpc = cmdRpcClient()

# send the command to the queue and wait for answer
response = command_rpc.call(args.command)

# output the response
print response

#==============================================================================+
# END OF FILE
#==============================================================================+
